### PR TITLE
Start documentation

### PR DESCRIPTION
**Pull request by Hermanverschooten**
Hey,
I made a start of documenting the public interface.

Offtopic question, would you mind keeping a change log?

Don't merge this yet, it still needs a lot of  work.

### CODE REVIEW

**Comment by Hermanverschooten** [comment]
If I understand the difference then it should be that the `bang` version raises when the text is too large to fit within the dimensions. But the return value of both is differente, while the ! version returns the `pid`, the normal version does not, necessitating an `elem(0)`.
Is this an oversight or a deliberate choice and if the latter, why and how do I document it?

**Comment by Hermanverschooten** [comment]
So I guess that completes the documentation, if you would please read it through and comment on any inaccuracies.

**Comment by Hermanverschooten** [comment]
Would you mind a name change of `Pdf.open/2` to `Pdf.run/2`? Open seems to suggest to me we are opening something, like a file, while we are actually running a function.

And can we change `Pdf.delete/1` to `Pdf.close/1` or `Pdf.finished/1`?  Delete seems to suggest we want to delete/remove perhaps the Pdf we just wrote to disk.

**Comment by andrewtimberlake** [comment]
> If I understand the difference then it should be that the `bang` version raises when the text is too large to fit within the dimensions. But the return value of both is differente, while the ! version returns the `pid`, the normal version does not, necessitating an `elem(0)`.
> Is this an oversight or a deliberate choice and if the latter, why and how do I document it?

The wrap functions need to return any content that doesn’t fit the wrap boundary so both `wrap` and `table` return `{pid, :complete}` or `{pid, remaining_content}`, because of this you can ignore any overflow with `elem(0)`.
If you are sure, but want the safety of a runtime error, you can use `wrap!` or `table!` which returns the `pid` or raises if there is content that doesn’t fit the supplied boundary.

The difference is in the chaining:
```
{pdf, :complete} = 
  pdf
  |> set_font("Helvetica", 10)
  |> Text.wrap({x, y}, {w, h}, "Some text")

{pdf, :complete} = 
  pdf
  |> Text.wrap({x2, y2}, {w2, h2}, "More text")

pdf
|> …
```
or
```
pdf
|> set_font("Helvetica", 10)
|> Text.wrap!({x, y}, {w, h}, "Some text")
|> Text.wrap!({x2, y2}, {w2, h2}, "More text")
|> …
```

**Comment by andrewtimberlake** [comment]
> Would you mind a name change of `Pdf.open/2` to `Pdf.run/2`? Open seems to suggest to me we are opening something, like a file, while we are actually running a function.
> 
> And can we change `Pdf.delete/1` to `Pdf.close/1` or `Pdf.finished/1`? Delete seems to suggest we want to delete/remove perhaps the Pdf we just wrote to disk.

I’m not sure that `Pdf.run/1` is correct. `Pdf.open/1` starts a gen_server.
I’m happy with `Pdf.close/1` rather than `Pdf.delete/1`
We could do `Pdf.start/1` but then neither `Pdf.close/1` nor `Pdf.delete/1` work.
`Pdf.delete/1` is only needed when handling multiple PDF documents in the same process, because if you don’t delete (clean up) the associated gen_server process, then you end up with massive memory leaks—I have been bitten by this when running a script to generate multiple PDF documents in a loop.
If you’re generating a PDF in a temporary process, like a web request then it’s not strictly necessary because all linked processes die with the process.

File: extra_doc/Tables.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+`Pdf.table!/5` will raise a `RuntimeError` if the data does not fit.
+
+## Options
+
+### General options
+These options can be applied to the table, rows and columns.
+
+Options | 
+:----- | :------
+`:background` | One of the predefined color atoms, see `Pdf.Color.color/1`
```
Comment (andrewtimberlake) [inline extra_doc/Tables.md]: Or an RGB tuple or a CMYK tuple, same as `set_fill_color`

File: extra_doc/Tables.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+`:border` | The thickness of the border.
+`:color` | The font color
```
Comment (andrewtimberlake) [inline extra_doc/Tables.md]: Also accepts the same color options as `:background`

**Comment by Hermanverschooten** [comment]
`open` does indeed start a `GenServer`, but it also stops again.
```elixir
  def open(opts \\ [], func) do
    {:ok, pdf} = new(opts)
    func.(pdf)
    delete(pdf)
    :ok
  end
```
So it doesn't really open anything, that's why I suggested `run`.

File: extra_doc/Tables.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+## Options
+
+### General options
+These options can be applied to the table, rows and columns.
+
+Options | 
+:----- | :------
+`:background` | One of the predefined color atoms, see `Pdf.Color.color/1`
+`:border` | The thickness of the border.
```
Comment (andrewtimberlake) [inline extra_doc/Tables.md]: Same as `:padding`

File: extra_doc/Tables.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+`:color` | The font color
+`:font_size` | Font size
+`:bold` | `true` or `false`
+`:italic` | `true` or `false`
+`:padding` | The padding to apply, integer for both or {horizontal, vertical}
```
Comment (andrewtimberlake) [inline extra_doc/Tables.md]: Same as CSS: `int` for all four sides, also accepts: `{int}`, `{horizontal, vertical}`, `{top, horizontal, bottom}`, `{top, right, bottom, left}`

File: extra_doc/Tables.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+`:colspan` | The number of columns to merge together
+
+### Table specific options
+
+Extra Options |  
+:------------ | :------
+`:cols` | Definitions for the individual columns.
+`:repeat_header` | The number of rows that make up the header of the table.
```
Comment (andrewtimberlake) [inline extra_doc/Tables.md]: Perhaps: "The number of rows to repeat when the data overflows the supplied boundary"

File: extra_doc/Tables.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+Extra Options | 
+:----- | :-----
+`:width` | Fixed width
+`:max_width` | The maximum column width
+`:min_width` | The minimum column width
+
+### Row specific options
+`:rows` takes a map of row numbers and a list of row definitions.
+Valid row numbers are `0..number of rows in data` and `-1` for the last row.
```
Comment (andrewtimberlake) [inline extra_doc/Tables.md]: It allows any number of negative indexes: -1 for last row, -2 for second last row, -3 for third last row etc.

**Review by andrewtimberlake — CHANGES_REQUESTED** [review]
Thanks for all this work. I have made some suggestions.

**Comment by andrewtimberlake** [comment]
> `open` does indeed start a `GenServer`, but it also stops again.
> …
> So it doesn't really open anything, that's why I suggested `run`.

You are right. I should have actually checked. I was thinking about `new/1` 🤦‍♂️

**Comment by andrewtimberlake** [comment]
We can also add the build indicator to the README:
[Build Status](https://travis-ci.org/andrewtimberlake/elixir-pdf)
`[Build Status](https://travis-ci.org/andrewtimberlake/elixir-pdf)`

File: extra_doc/Tables.md (hunk 5 above)
Comment (Hermanverschooten) [reply extra_doc/Tables.md]: maybe just add: ",and that will be repeated when the data overflows the supplied boundary."

**Review by andrewtimberlake — APPROVED** [review]

**Pull request merged**